### DISCUSSION TITLE
rust-analyzer is adding some `target` folders to `didChangeWatchFiles` watchers

### DISCUSSION THREAD

**Discussion by alex-ds13** [mrcjkb/rustaceanvim#261] [Q&A] [answered]
### Neovim version (nvim -v)

v0.10.0-dev

### Operating system/version

Windows 11

### Output of :checkhealth rustaceanvim

```console
==============================================================================
rustaceanvim: require("rustaceanvim.health").check()

Checking for Lua dependencies ~
- WARNING dap not installed. Needed for debugging features [mfussenegger/nvim-dap](https://github.com/mfussenegger/nvim-dap)

Checking external dependencies ~
- OK rust-analyzer: found rust-analyzer 0.3.1839-standalone (c06ca6cff 2024-02-11)
- OK Cargo: found 
- OK rustc: found 

Checking config ~
- OK No errors found in config.

Checking for conflicting plugins ~
- OK No conflicting plugins detected.

Checking for tree-sitter parser ~
- OK tree-sitter parser for Rust detected.
```

### How to reproduce the issue

I'm not sure how to reproduce this. When I have a rust-analyzer LSP running and at the same time do anything with cargo, like `cargo update` or `cargo build` I get a lot of `EPERM` errors like this:

```console
Error executing luv callback:
...\neovim\nvim-win64\share\nvim\runtime/lua/vim/_watch.lua:68: EPERM: operation not permitted: c:/Projects/Other/Rust/b
udgetLookup/target/release/deps/libunicode_ident-1737a7a6bd2cc11e.rlib.temp-archive-536a548.a
stack traceback:
        [C]: in function 'assert'
        ...\neovim\nvim-win64\share\nvim\runtime/lua/vim/_watch.lua:68: in function <...\neovim\nvim-win64\share\nvim\ru
ntime/lua/vim/_watch.lua:55>
Press ENTER or type command to continue
Error executing luv callback:
...\neovim\nvim-win64\share\nvim\runtime/lua/vim/_watch.lua:68: EPERM: operation not permitted: c:/Projects/Other/Rust/b
udgetLookup/target/release/deps/libunicode_ident-1737a7a6bd2cc11e.rlib.temp-archive-536a548.a
stack traceback:
        [C]: in function 'assert'
        ...\neovim\nvim-win64\share\nvim\runtime/lua/vim/_watch.lua:68: in function <...\neovim\nvim-win64\share\nvim\ru
ntime/lua/vim/_watch.lua:55>
Press ENTER or type command to continue
```
From one of the project I have, I've noticed that the rust LSP client has the following:
```
dynamic_capabilities = {
      capabilities = {
        ["workspace/didChangeWatchedFiles"] = { {
            id = "workspace/didChangeWatchedFiles",
            method = "workspace/didChangeWatchedFiles",
            registerOptions = {
              watchers = { {
                  globPattern = "C:\\Projects\\Other\\Rust\\budgetLookup/**/*.rs"
                }, {
                  globPattern = "C:\\Projects\\Other\\Rust\\budgetLookup/**/Cargo.toml"
                }, {
                  globPattern = "C:\\Projects\\Other\\Rust\\budgetLookup/**/Cargo.lock"
                }, {
                  globPattern = "C:\\Projects\\Other\\Rust\\budgetLookup\\target\\x86_64-pc-windows-msvc\\debug\\build\\
budget_lookup-a7939d74eaadce18\\out/**/*.rs"
                }, {
                  globPattern = "C:\\Projects\\Other\\Rust\\budgetLookup\\target\\x86_64-pc-windows-msvc\\debug\\build\\
budget_lookup-a7939d74eaadce18\\out/**/Cargo.toml"
                }, {
                  globPattern = "C:\\Projects\\Other\\Rust\\budgetLookup\\target\\x86_64-pc-windows-msvc\\debug\\build\\
budget_lookup-a7939d74eaadce18\\out/**/Cargo.lock"
                } }
            }
          } }
      },
```

It adds some `target` folder to the watchers which makes neovim try to watch those files, but if I'm running the binary or doing a `cargo build` those files will be under use and nvim will fail to access them with the `EPERM` error.

I'm not sure what makes this happen, from what I've tried to search there doesn't seem to be much about these types of errors. I know this might not be a problem with rustaceanvim but since you use this on neovim I was wondering if you or anyone else already got something like this and know how to fix it.

### Expected behaviour

No errors to show up.

### Actual behaviour

A lot of `EPERM` errors show all the time.

### The minimal config used to reproduce this issue.

Again no minimal config to reproduce, this happens with every project of rust I use on neovim. Except the folders from `target` showing on the watchers, that has happened only on one project and I can't seem to either reproduce it or get rid of it.

**Comment by mrcjkb** [comment]
Hey 👋 

The `didChangeWatchedFiles` client capability is an experimental feature that was introduced (disabled by default) in Neovim 0.9.
It has recently been enabled by default in https://github.com/neovim/neovim/pull/23190.

The feature still has some issues, which need ironing out. See https://github.com/neovim/neovim/issues/23291, which is probably related to your problem.

I haven't observed the behaviour you describe on Linux, so yours might be specific to Windows. Might be worth dropping a comment in the neovim issue.

To work around it, I suggest you either switch back to neovim 0.9 or disable the capability (as described in the neovim issue).

**Reply by alex-ds13** [reply]
Yes, I've already created a feature request on https://github.com/neovim/neovim/issues/27646 to allow setting the excluded folders, thinking it could fix this but I tried it myself and it still didn't fix it because it will give priority to the folders set as watchers by the LSP server and it was only then that I noticed that rust-analyzer had those `target` folders as watchers. From what I can see rustaceanvim doesn't do anything with the `didChangeWatchedFiles` capability, so it has to be rust-analyzer itself that is adding these folders, but I don't know why.

**Comment by mrcjkb** [comment, marked as answer]
For documentation: This is fixed with https://github.com/neovim/neovim/pull/27347
